File: 206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        track = []
        #add all values in linked list
        while head:
            track.append(head.val)
            head = head.next

        dummy = ListNode()
        current = dummy
        
        while track:
            current.next = ListNode(track.pop())
            current = current.next

        return dummy.next

File: 206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {

        //change pointers direction
        ListNode* previous = nullptr;
        ListNode* current = head;

        //iterate till current is empty
        while(current){
            ListNode* temp = current->next;
            current->next = previous;
            previous = current;
            current = temp;
        }

        //previous becomes the new head
        return previous;

        //T O(n) S O(1)
        // most optimal
    }
};